doc/reload: Expand rule-reload discussion (#13522)
OISF/suricata: doc/userguide/rule-management/rule-reload.rst

Rule Reloads
============

Suricata was designed to reload rules while it is actively processing
network traffic to minimize service disruption.

Suricata must be administratively directed to reload rules while it is running; there
are two ways to cause this:

* Via a UNIX domain socket with the command ``reload-rules`` or ``ruleset-reload-nonblocking``
* Upon receipt of the ``USR2`` signal

Suricata performs these steps while reloading rules:

* The main Suricata configuration is reloaded to update rule variables and values; this
  includes the rule related files ``classification.config``, ``reference.config`` and
  ``theshold.config``
* All rule files are reloaded with new rule variables applied
* A new detection engine is created for the updated rules.
* The previous and newly created detection engines are swapped.
* Ensure all threads are updated

> * Free old detection engine and associated resources.

Reviewer: This looks devguide worthy to me. Not so sure if this is of relevance to the user. wdyt?

Reviewer 2: I think something is useful here for end-users, like how long and the is the old detection engine freed? At the end of the flows?  Immediately when the new one is ready?

Reviewer 3: Maybe we could either add it to the Devguide (somewhere under `Suricata Internals), _or_ have this as subsection here (maybe the last one?), with a tittle like `Advanced: Reload Steps` -- as this can be useful to understand changes in performance, for instance, when that command is invoked?

Author: Ill make it a subsection.

Suricata will continue to process packets during the update process. Note that additional system
memory is used during the reload process as a new detection engine and the reloaded rules are
associated with it.

Reload Triggers
~~~~~~~~~~~~~~~

Via process signal::

  kill -USR2 $(pidof suricata)

Via the UNIX domain socket.

Blocking reload ::

  suricatasc -c reload-rules

Non-blocking reload ::

  suricatasc -c ruleset-reload-nonblocking

It is also possible to get information about the last reload via dedicated commands.
See :ref:`standard-unix-socket-commands` for more information.

Resources Reloaded
~~~~~~~~~~~~~~~~~~

There are two types of resources that are reloaded during a rule reload

* Rule related configuration:
  * Suricata's configuration file(s): ``suricata.yaml`` and any specified with the command-line

>     options ``--include <config-file.yaml>``

Reviewer: Q: Is all of the configuration file `suricata.yaml` reloaded? There seem to be parts that'd be hard to reload while the engine is still up and running.

Reviewer 2: I think we need an itemized list of all that is reloaded. However, I thought we already had that somewhere, but can't find it.

Author: The configuration file is reloaded -- however, not all values take effect.

The items in `vars` (address and port groups) take effect as rules are loaded.
Items in related files -- {threshold,reference,classification}.config take effect.

I'm not sure how to identify the other loaded values that take effect.

  * Ancillary rule-related configuration files: ``classification.config``, ``reference.config``
  and ``threshold.config``

> Situations to reload

Author: Good idea.

~~~~~~~~~~~~~~~~~~~~~

Rule reloads are used in situations when:

* Rules have been changed since the last reload. Vendors often add rules frequently and
  sometimes update existing rules. Rules should be reloaded according to a security policy
  that includes Suricata content updates.

> * Rule variables have been changed. Rule reloads will use rule variables from the Suricata
>   configuration file. Rule variable changes should be followed by reloading rules which
>   will cause the updated rule variables to take effect.

Reviewer 3: Possible suggestion...
```suggestion
* Rule variables have been changed. Rule reloads will use rule variables from the Suricata
  configuration file. When updating these, reload the rules in order to the updated rule variables take effect.
```
Or something like that?
